dataset/dataset.cpp

```diff
@@ -121,47 +121,27 @@ namespace extents {
 scipp::index makeUnknownEdgeState(const scipp::index extent) {
   return -extent - 1;
 }
-scipp::index shrink(const scipp::index extent) { return extent - 1; }
 bool isUnknownEdgeState(const scipp::index extent) { return extent < 0; }
-scipp::index decodeExtent(const scipp::index extent) {
+scipp::index decode(const scipp::index extent) {
   if (isUnknownEdgeState(extent))
     return -extent - 1;
   return extent;
 }
-bool isSame(const scipp::index extent, const scipp::index reference) {
-  return reference == -extent - 1;
-}
-bool oneLarger(const scipp::index extent, const scipp::index reference) {
-  return extent == -reference - 1 + 1;
-}
-bool oneSmaller(const scipp::index extent, const scipp::index reference) {
-  return extent == -reference - 1 - 1;
-}
 void setExtent(std::unordered_map<Dim, scipp::index> &dims, const Dim dim,
                const scipp::index extent, const bool isCoord) {
   const auto it = dims.find(dim);
-  // Internally use negative extent -1 to indicate unknown edge state. The `-1`
-  // is required for dimensions with extent 0.
-
   if (it == dims.end()) {
-    dims[dim] = extents::makeUnknownEdgeState(extent);
-  } else {
-    auto &heldExtent = it->second;
-    if (extents::isUnknownEdgeState(heldExtent)) {
-      if (extents::isSame(extent, heldExtent)) { // Do nothing
-      } else if (extents::oneLarger(extent, heldExtent) && isCoord) {
-        heldExtent = extents::shrink(extent);
-      } else if (extents::oneSmaller(extent, heldExtent) && !isCoord) {
-        heldExtent = extent;
-      } else {
-        throw except::DimensionError(heldExtent, extent);
-      }
-    } else {
-      // Check for known edge state
-      if ((extent != heldExtent || isCoord) && extent != heldExtent + 1)
-        throw except::DimensionError(heldExtent, extent);
-    }
+    dims[dim] = isCoord ? extents::makeUnknownEdgeState(extent) : extent;
```

Comment:
No not a bug, this is related to the behavior change: Previously we also supported masks or data on the edges.

Actually I think testing this in isolation may be counter-productive. We do not care about the extents logic, but about the resulting behavior of `DataArray`. Testing implementation details will lock is more into a particular implementation, rather than helping us to test the actual required concepts/contracts of the data structures (dataset and data array) that rely on it.